src/Workspaces/Remote/Core/ServiceDescriptors.cs

```diff
@@ -40,28 +40,25 @@ internal static class ServiceDescriptors
         private const string InterfaceNamePrefix = "IRemote";
         private const string InterfaceNameSuffix = "Service";
 
-        // CONSIDER: The descriptors with callbacks currently depend on the feature implementations.
-        // To avoid this coupling we could moe the map from service type to callback dispatcher to IRemoteHostClientProvider and populate it via MEF.
-
         internal static readonly ImmutableDictionary<Type, (ServiceDescriptor descriptor32, ServiceDescriptor descriptor64)> Descriptors = ImmutableDictionary.CreateRange(new[]
         {
             CreateDescriptors(typeof(IRemoteAssetSynchronizationService)),
             CreateDescriptors(typeof(IRemoteAsynchronousOperationListenerService)),
-            CreateDescriptors(typeof(IRemoteTodoCommentsDiscoveryService), callbackInterface: typeof(IRemoteTodoCommentsDiscoveryService.ICallback), new RemoteTodoCommentsDiscoveryCallbackDispatcher()),
-            CreateDescriptors(typeof(IRemoteDesignerAttributeDiscoveryService), callbackInterface: typeof(IRemoteDesignerAttributeDiscoveryService.ICallback), new RemoteDesignerAttributeDiscoveryCallbackDispatcher()),
-            CreateDescriptors(typeof(IRemoteProjectTelemetryService), callbackInterface: typeof(IRemoteProjectTelemetryService.ICallback), new RemoteProjectTelemetryServiceCallbackDispatcher()),
+            CreateDescriptors(typeof(IRemoteTodoCommentsDiscoveryService), callbackInterface: typeof(IRemoteTodoCommentsDiscoveryService.ICallback)),
```

Comment:
💭 It seems like the second argument here shouldn't be necessary.

Reply:
It is necessary since the descriptor is needed in service factories, where MEF composition is not available.

Reply:
Essentially, the descriptors define contract. The dispatchers implement the contract (the callback part, that is).